### ISSUE TITLE
MongoDB Storage/Caching

### ISSUE THREAD

**Issue by adunna** [AHAAAAAAA/PokemonGo-Map#588]
# New Feature - MongoDB Storage/Caching

Change from previous SQLite implementation to faster Mongo-based storage for caching.

Checklist:
- [ ] Implement base MongoDB storage
- [ ] Store Pokestop data
- [ ] Pokestop lure check refresh, new Pokestop check refresh, & update DB
- [ ] Store Gym data
- [ ] Gym team, leader, Pokemon, prestige, etc. check refresh, new Gym check refresh, & update DB
- [ ] Store Pokemon data
- [ ] Pokemon check refresh & update DB

Referencing [Pull #264](https://github.com/AHAAAAAAA/PokemonGo-Map/pull/264)

**Comment by rezz3x** [comment]
would be nice if you could decide whatever storage you want to use, and mysql implementation would be pretty sick too

**Comment by scarecr0w12** [comment]
Would be nice to have the option ability, But yes, MySQL would Definitely be a big one to have functionality with.

**Comment by mantas7776** [comment]
Please dont use mongoDB, it loses data.
Source: http://cryto.net/~joepie91/blog/2015/07/19/why-you-should-never-ever-ever-use-mongodb/

**Comment by adunna** [comment]
That article is over a year old - please post a more recent article working off the latest Mongo version.

**Comment by mkody** [comment]
Anyway, give the user the choice to decide which SGBD it wants to use.

**Comment by adunna** [comment]
I'll be implementing Mongo first since it is the easiest for me to deploy for tests - but yes, will eventually implement other solutions, secondarily MySQL.

**Comment by dorian-marchal** [comment]
Why isn't SQLite enough ?

**Comment by mkody** [comment]
Yeah, now that I'm thinking of it, I don't see why we should do this. See https://www.sqlite.org/whentouse.html

**Comment by adunna** [comment]
Couple of reasons. First is for the write capabilities, as SQLite is limited to one write at a time - we do not know how much traffic we will get, nor how much we will be writing although with the way the API is going, it's best to try and cache all that we can. So, we can expect a lot of writing and a lot of users needing to do a lot of writing. To keep speed, it's best to allow that to be done concurrently. Second is that depending on how the mobile applications go, it could be setup in a client-server format for the mobile client to communicate with our server for caching, in which case we'd need a client/server engine such as Mongo or MySQL.

**Comment by clbuttic** [comment]
A centralized caching service will help alleviate some of the load on the servers. Anyone keen on kicking something off? Starting off with a simple replication of the sqllite db.

**Comment by alvinycheung** [comment]
I'm building a node server that this can talk to to write data to a mysql database.

**Comment by alvinycheung** [comment]
I think a centralized caching service is a great idea! So, I'm building a node server that this can talk to to write data to a mysql database. It should be pretty painless https://github.com/AHAAAAAAA/PokemonGo-Map/issues/739

**Comment by MikeDX** [comment]
Is mongo the one that requires 8gb before you even start?

Also I'm working on some code to only return the data required within the viewport but this is on hold until another part of the search code has been merged. I've got the branch ready to push, but I can't do anything with it until the code for deduplicating the pokestops/gyms has been merged with develop

**Comment by arthurkay** [comment]
Yes, mongo is the one that will refuse to start without several gigabytes of free space unless you specify a command line argument.

Honestly its fine. Since it would be managed by the app that flag would be used.

External mongo instances would obviously be set up by the user.

**Comment by lisandro52** [comment]
That MongoDB that doesn't return all the results from queries?
https://engineering.meteor.com/mongodb-queries-dont-always-return-all-matching-documents-654b6594a827
RethinkDB/CouchDB would be better choices if you want to go the NoSQL way.
